[functorch] add switch to enable autograd.Function

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88785
* __->__ #88784

This is mostly a debug or "if you know what you're doing" switch for
now. It is not public API.

Test Plan:
- new tests

zou3519 committed:
[functorch] add switch to enable autograd.Function

This is mostly a debug or "if you know what you're doing" switch for
now. It is not public API.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[functorch] add switch to enable autograd.Function"


This is mostly a debug or "if you know what you're doing" switch for
now. It is not public API.

Test Plan:
- new tests

[ghstack-poisoned]

samdow commented on aten/src/ATen/functorch/DynamicLayer.cpp:
```diff
@@ -83,6 +83,18 @@ static DynmetaData& getGlobalDynmetaData() {
 }
 
+namespace {
+thread_local bool kIsAutogradFunctionAllowed = false;
```
I would expect that if I enabled autograd.Function in one thread, all spawned threads would also have autograd.Function enabled (not certain if this will come up, but thinking that it might because of multithreading of the backward pass combined with higher order gradients). This version means that all threads will end up starting with the false behavior, right?

zou3519 replied:
You're right, ATen/ThreadLocalState needs to understand this. Thanks for catching

samdow commented on aten/src/ATen/functorch/DynamicLayer.h:
```diff
@@ -113,6 +113,12 @@ TORCH_API Tensor unwrapIfDead(const Tensor& tensor);
 TORCH_API std::ostream& operator<<(std::ostream& os, const DynamicLayer& layer);
 TORCH_API std::ostream& operator<<(std::ostream& os, const std::vector<DynamicLayer>& dynamicLayerStack);
 
+// While a functorch grad transform is active, autograd.Function is disabled
```
nit: Isn't it all functorch transforms? Since vmap is a problem if the input tensor requires grad

zou3519 replied:
Yup, you're right. Will update

zou3519 committed:
Update on "[functorch] add switch to enable autograd.Function"


This is mostly a debug or "if you know what you're doing" switch for
now. It is not public API.

Test Plan:
- new tests

[ghstack-poisoned]

samdow reviewed (approved):
Awesome thanks for the updates!

soulitzer commented on test/functorch/test_eager_transforms.py:
```diff
@@ -2388,6 +2388,35 @@ def f(x):
+                x, = ctx.saved_tensors
+                return gy * x.cos()
+
+        x = torch.randn([])
```
Is this tensor intentionally doesn't require gradients? I guess it doesn't matter because should error anyway.

zou3519 replied:
The functorch grad transform doesn't use the requires_gradness of the tensor to compute gradients, it accepts an argnums argument that says which input to compute gradients with respect to

soulitzer replied:
Oh whoops, good point.
